File: 2016f/soln/NewtonsNightmare.py
import math

def target_angle(layers, target_x):
    angle = 0.0
    #STUBIFY
    l = math.pi * 0.5
    r = 0.0
    EPSILON = 1e-9
    while math.fabs(l - r) > EPSILON:
        angle = 0.5 * (l + r)
        incidence_angle = 0.5 * math.pi - angle
        prev_refractive_index = layers[0][1]
        x = math.tan(incidence_angle) * layers[0][0]
        other_condition = False
        for height, refractive_index in layers[1:]:
            sin_angle = math.sin(incidence_angle) * prev_refractive_index / refractive_index
            if sin_angle > 1 or sin_angle < 0:
                r = angle
                other_condition = True
                break
            incidence_angle = math.asin(sin_angle)
            prev_refractive_index = refractive_index
            x += math.tan(incidence_angle) * height
        if not other_condition:
            if x > target_x:
                r = angle
            else:
                l = angle
    angle = angle * (180.0 / math.pi)
    #ENDSTUBIFY
    return round(angle, 4)

if __name__ == "__main__":
    with open("NewtonsNightmareIN.txt", "r") as f:
        while True:
            s = f.readline()
            if s == '':
                break
            N, X = [int(x) for x in s.split()]
            layers = []
            for i in range(N):
                s = f.readline().split()
                layers.append((float(s[0]), float(s[1])))
            print("%.4f" % target_angle(layers, X))

File: 2016f/soln/NewtonsNightmare.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.Scanner;

public class NewtonsNightmare {

    public static void main(String[] args) throws FileNotFoundException {
        // Before submitting, make sure the main method hasn't been changed!
        Scanner sc = new Scanner(new FileReader("NewtonsNightmareIN.txt"));
        while(sc.hasNext()) {
            String[] firstLine = sc.nextLine().split(" ");
            int numLines = Integer.parseInt(firstLine[0]);
            double targetX = Double.parseDouble(firstLine[1]);
            
            int[] heights = new int[numLines];
            double[] indexes = new double[numLines];
            
            for(int i = 0; i < numLines; i++) {
            	String[] curLine = sc.nextLine().split(" ");
            	heights[i] = Integer.parseInt(curLine[0]);
            	indexes[i] = Double.parseDouble(curLine[1]);
            }
            
            System.out.printf("%.4f\n", findAngle(heights, indexes, targetX));
        }
        sc.close();
    }
    
    private static double findAngle(int[] heights, double[] indexes, double targetX) {
    	double angle = 0.0;
    	double l = Math.PI * 0.5;
    	double r = 0.0;
    	double maxError = .000000001;
    	while (Math.abs(l - r) > maxError) {
    		angle = 0.5 * (l + r);
    		double incidenceAngle = 0.5 * Math.PI - angle;
    		double prevIndex = indexes[0];
    		double x = Math.tan(incidenceAngle) * (double)heights[0];
    		boolean otherCondition = false;
    		for (int i = 1; i < heights.length; i++) {
    			double sinAngle = Math.sin(incidenceAngle) * prevIndex / indexes[i];
    			if ((sinAngle > 1) || (sinAngle < 0)) {
    				r = angle;
    				otherCondition = true;
    				break;
    			}
    			incidenceAngle = Math.asin(sinAngle);
    			prevIndex = indexes[i];
    			x += Math.tan(incidenceAngle) * (double)heights[i];
    		}
    		if (!otherCondition) {
    			if (x > targetX) {
    				r = angle;
    			} else {
    				l = angle;
    			}
    		}
    	}
    	angle = angle * (180.0 / Math.PI);
    	return Math.round(angle * 10000.0)/10000.0;
    	
    }
}
